### PR TITLE
Poc/command blocks

### PR DESCRIPTION
**Pull request by ncallaway**
This is simply a proof of concept. It can be demo code for a starting point, but there's still some more work to do.

Things remaining:
- Adding a status to the result object
- Command result execution matching the desired status
- Perform "was my result block fully satisfied" _before_ executing the result block
- Don't require the result to be fully satisfied if no command block was provided
- Code cleanup

### CODE REVIEW

File: test/lib/command_test.rb [hunk 1]
```diff
@@ -27,46 +27,78 @@ def test_executes_block_passed_to_command
...
+  end
+
   def test_result_and_command_available_in_block
     NoArgCommand.call do |r, c|
       assert c.is_a? NoArgCommand
 
       assert r.is_a? Slayer::Result
       assert_equal true, r.success?
+
+      r.pass; r.fail
```
Comment (wkirby) [inline test/lib/command_test.rb]: Should we include a third possibility, `r.skip`, which allows you to do nothing with the response object?

File: lib/slayer/result.rb [hunk 2]
```diff
@@ -1,12 +1,34 @@
...
+    def pass(for_status = nil, &block)
+      @handled_pass ||= for_status == nil || for_status == :default
+
+      yield if (block_given? && success? && (for_status == nil || for_status == :default || for_status == @result))
+    end
+
+    def fail(for_status = nil, &block)
+      @handled_fail ||= for_status == nil || for_status == :default
+
+      yield if (block_given? && failure? && (for_status == nil || for_status == :default || for_status == @result))
```
Comment (wkirby) [inline lib/slayer/result.rb]: @houndci-bot is being a pain here, but I do agree that this yield conditional is a bit hard to parse.

**Comment by wkirby** [comment]
Two things:

1. I think we should add a method for `r.skip` which will let you skip handling of the result object. I can't think of a specific use case in our code, but at least you're being explicit that you don't want to handle the result and you don't have to call `r.pass; r.fail`.
2. I think the `Slayer::Result` object still has a `result` attribute, which should really be changed to something like `payload` so we're not looking at `r.result`.

File: test/lib/command_test.rb (hunk 1 above)
Comment (ncallaway) [inline test/lib/command_test.rb]: That makes sense to me, although I'm not sure I see the use-case for the block at all in that case.

If you're going to `r.skip`, then you shouldn't need to pass a block at all. I think the impl should be improved to not mandate a block. If no block is provided, you still get the result object back from the call. The _only_ real reason for the block, in my thoughts, was to require both the pass and the fail check.

File: lib/slayer/result.rb (hunk 2 above)
Comment (ncallaway) [inline lib/slayer/result.rb]: Oh yea, for sure. This was proof of concept, and I absolutely wouldn't recommend merging this pull request as is.

**Comment by ncallaway** [comment]
For 1. I commented inline. I'm happy to add it, although I think the "just don't pass a block" is a little bit of a cleaner resolution to that use-case.

For 2. I would recommend a `Slayer::Result` has `value` which contains the return value, `status` which contains a more detailed machine interpretable status code than just pass or fail (e.g. `status` could be `:404` or `:500`), and `message` which contains a user presentable string explaining the error scenario in more detail.
